plugin-localization/src/main/java/com.mapbox.mapboxsdk.plugins.localization/LocalizationPlugin.java

```diff
@@ -0,0 +1,234 @@
+      || languageTag.equals("zh-Hans");
+  }
+
+  /**
+   * Retrieves and adjusts the individual map layers that need to be adjusted
+   *
+   * @param languageToSetMapTo the language that the map is being adjusted to
+   */
+  private void adjustMapTextLayers(String languageToSetMapTo) {
+    for (Source source : mapboxMap.getSources()) {
+      if (sourceIsFromMapbox(source)) {
+        for (Layer layer : mapboxMap.getLayers()) {
+          if (layerHasAdjustableTextField(layer)) {
+            if (((SymbolLayer) layer).getTextField().getValue().contains("{name")
+              || !getDeviceLanguage().equals("en") && ((SymbolLayer) layer).getTextField().getValue().contains("{abbr}")
+              ) {
+              layer.setProperties(textField(String.format("{name_%s}", languageToSetMapTo)));
+            }
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Sets map text to a specified language. If the language is not supported by OpenStreetMap, then the
+   * default language (specified during the instantiation of the plugin) is checked for OSM support. If
+   * it is supported by OSM, then the map will be changed to the default language. English is the last-resort
+   * language used in case the default language isn't supported by OSM neither.
+   *
+   * @param language The language that to set the map text to.
+   * @since 0.1.0
+   */
+  public void setMapTextLanguage(String language) {
+    if (mapboxSourceSupportsLanguage(language)) {
+      adjustMapTextLayers(language);
+    } else if (mapboxSourceSupportsLanguage(selectedBackupLanguage)) {
+      adjustMapTextLayers(selectedBackupLanguage);
+    } else {
+      adjustMapTextLayers("en");
+    }
+  }
+
+  /**
+   * Checks whether the map's source is a source provided by Mapbox, rather than a custom source.
+   *
+   * @param singleSource an individual source object from the map
+   * @return true if the source is from the Mapbox Streets vector source, false if it's not.
+   */
+  private boolean sourceIsFromMapbox(Source singleSource) {
+    return singleSource instanceof VectorSource
+      && ((VectorSource) singleSource).getUrl().substring(0, 9).equals("mapbox://")
+      && (((VectorSource) singleSource).getUrl().contains("mapbox.mapbox-streets-v7")
+      || ((VectorSource) singleSource).getUrl().contains("mapbox.mapbox-streets-v6"));
+  }
+
+  /**
+   * Checks whether a single map layer has a textField that could potentially be localized to the device's language.
+   *
+   * @param singleLayer an individual layer from the map
+   * @return true if the layer has a textField eligible for translation, false if not.
+   */
+  private boolean layerHasAdjustableTextField(Layer singleLayer) {
+    return singleLayer instanceof SymbolLayer && (((SymbolLayer) singleLayer).getTextField() != null
+      && (((SymbolLayer) singleLayer).getTextField().getValue() != null
+      && !(((SymbolLayer) singleLayer).getTextField().getValue().isEmpty())));
```

Comment:
As tail work, consider making this plugin also affect layers whose `text-field` property is set to a function. In the Mapbox Streets style, for example, U.S. state labels are a function with stops for either `{name_en}` or `{abbr}` based on the zoom level.

Without accounting for functions, there will be some inconsistency in state and country labels at low zoom levels. However, expressions will be available in the next release, so it might be worth spending much effort accommodating functions in the meantime.

/ref mapbox/mapbox-gl-native#10713 mapbox/mapbox-gl-native#10944

Reply:
> However, expressions will be available in the next release, so it might be worth spending much effort accommodating functions in the meantime.

I'd prefer merging this PR and fixing this in a different branch in the future when we switch to 6.0.0.

Reply:
@cammace Is there a ticket in this repo where we're tracking this tail work?

Reply:
I ticketed this out in https://github.com/mapbox/mapbox-plugins-android/issues/301